design/design-doc-composition.md

```diff
@@ -0,0 +1,1079 @@
+
+* Often supporting resources must reference the managed resource they support,
+  for example an Azure `kind: MySQLServerVirtualNetworkRule` must reference the
+  `kind: MySQLServer` it applies to. Dynamically provisioned managed resources
+  such as a `kind: MySQLServer` have non-deterministic names, making it
+  impossible to create a `kind: MySQLServerVirtualNetworkRule` until the `kind:
+  MySQLServer` it must reference has been provisioned.
+* When a resource class references a statically provisioned managed resource
+  every managed resource that is dynamically provisioned using that class will
+  reference that specific managed resource. For example if a `kind:
+  GKEClusterClass` references a `kind: Subnetwork` then every `kind: GKECluster`
+  dynamically provisioned using said class will attempt to share said `kind:
+  Subnetwork`, despite it often being desirable to create a unique `kind:
+  Subnetwork` for each dynamically provisioned `kind: GKECluster`.
+
+The one-to-one relationship between resource claims and resource classes thus
+weakens portability, separation of concerns, and support for [GitOps]. An
+infrastructure operator can publish a resource class representing a single
+managed resource that an application operator may dynamically provision, but in
+the likely event that managed resource requires supporting managed resources to
+function usefully the application operator must ask an infrastructure operator
+to provision them.
+
+Furthermore, defining a core set of portable resource claims has begun to limit
+Crossplane. Resource claims are subject to the [lowest common denominator]
+problem; when a claim may provide configuration inputs that may be used to match
+or provision many kinds of managed resource it may support only the settings
+that apply to _all_ compatible managed resources. This is in part why Crossplane
+defines relatively few resource claim kinds. Meanwhile, it's possible that an
+infrastructure operator deploys Crossplane to provide an opinionated abstraction
+for the application operators in their organisation and that said organisation
+only uses AWS. If this organisation values Crossplane's separation of concerns
+but does not need its portability there is no reason that its application
+operators should be limited to resource claims that are constrained by
+supporting all possible providers.
+
+## Goals
+
+The goal of this document is to propose a simple, consistent, Kubernetes native
+model for composing applications and their infrastructure needs.
+
+TODO(negz): Elaborate on the below goals.
+
+* Nested composition
+* User defined abstractions
+* Shareable abstractions
+* Composable infrastructure
+* Discoverable abstraction to composition relationships
+* Compositions configured inside API boundary
+
+## Use Cases
+
+The proposal put forward by this document is intended to address a handful of
+use cases around application and infrastructure provisioning.
+
+### Application Composition
```

Comment:
Again, I wonder how different is the application composition from the appConfig?

Reply:
it's different in the sense that the composition itself becomes a new type/CRD in the system, enabling reuse at the API level (not source/yaml level).

Reply:
Sounds like it can be part of OAM in the sense that a trait/workload can be a composition itself?

Reply:
in OAM speak, imagine if every `ApplicationConfiguration` was associated with a CRD representing it's input schema, and could be component in another `ApplicationConfiguration`.